IOMC/ParticleGuns/src/BeamMomentumGunProducer.cc

```diff
@@ -0,0 +1,156 @@
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/RandomNumberGenerator.h"
+
+#include "CLHEP/Random/RandFlat.h"
+
+#include "TFile.h"
+
+namespace CLHEP {
+  class HepRandomEngine;
+}
+
+namespace edm {
+  BeamMomentumGunProducer::BeamMomentumGunProducer(const edm::ParameterSet& pset)
+      : FlatBaseThetaGunProducer(pset),
+        parPDGId_(nullptr),
+        parX_(nullptr),
+        parY_(nullptr),
+        parZ_(nullptr),
+        parPx_(nullptr),
+        parPy_(nullptr),
+        parPz_(nullptr),
+        b_npar_(nullptr),
+        b_eventId_(nullptr),
+        b_parPDGId_(nullptr),
+        b_parX_(nullptr),
+        b_parY_(nullptr),
+        b_parZ_(nullptr),
+        b_parPx_(nullptr),
+        b_parPy_(nullptr),
+        b_parPz_(nullptr) {
+    edm::ParameterSet pgun_params = pset.getParameter<edm::ParameterSet>("PGunParameters");
+
+    // doesn't seem necessary to check if pset is empty - if this
+    // is the case, default values will be taken for params
+    zpos_ = pgun_params.getParameter<double>("ZPosition");
+    edm::FileInPath fp = pgun_params.getParameter<edm::FileInPath>("FileName");
+    std::string infileName = fp.fullPath();
+
+    fFile_ = new TFile(infileName.c_str());
+    fFile_->GetObject("EventTree", fTree_);
+    nentries_ = fTree_->GetEntriesFast();
+    if (fVerbosity > 0)
+      edm::LogVerbatim("BeamMomentumGun") << "Total Events: " << nentries_ << " in " << infileName << " Z " << zpos_;
+
+    rand_.SetSeed(0);
+
+    // Set branch addresses and branch pointers
+    fTree_->SetBranchAddress("npar", &npar_, &b_npar_);
+    fTree_->SetBranchAddress("eventId", &eventId_, &b_eventId_);
+    fTree_->SetBranchAddress("parPDGId", &parPDGId_, &b_parPDGId_);
+    fTree_->SetBranchAddress("parX", &parX_, &b_parX_);
+    fTree_->SetBranchAddress("parY", &parY_, &b_parY_);
+    fTree_->SetBranchAddress("parZ", &parZ_, &b_parZ_);
+    fTree_->SetBranchAddress("parPx", &parPx_, &b_parPx_);
+    fTree_->SetBranchAddress("parPy", &parPy_, &b_parPy_);
+    fTree_->SetBranchAddress("parPz", &parPz_, &b_parPz_);
+
+    produces<HepMCProduct>("unsmeared");
+    produces<GenEventInfoProduct>();
+
+    if (fVerbosity > 0)
+      edm::LogVerbatim("BeamMomentumGun") << "BeamMonetumGun is initialzed";
+  }
+
+  void BeamMomentumGunProducer::produce(edm::Event& e, const edm::EventSetup& es) {
+    if (fVerbosity > 0)
+      edm::LogVerbatim("BeamMomentumGun") << "BeamMomentumGunProducer : Begin New Event Generation";
+
+    edm::Service<edm::RandomNumberGenerator> rng;
+    CLHEP::HepRandomEngine* engine = &rng->getEngine(e.streamID());
+
+    // event loop (well, another step in it...)
+    // no need to clean up GenEvent memory - done in HepMCProduct
+    // here re-create fEvt (memory)
+    //
+    fEvt = new HepMC::GenEvent();
+
+    // random entry generation for peaking event randomly from tree
+    long int rjentry = rand_.Uniform(0, nentries_ - 1);
```

Comment:
@bsunanda I do not see reasons not to use the CLHEP uniform random number generator based on the engine from the service. The code as it is now breaks our usual rule of relying on random numbers managed by the FWK service, which makes them reproducible.

Reply:
I am now using the random # generator from CLHEP. Sorry that I overlooked this

Reply:
@bsunanda thank you a lot